[Daniel Lee] iP

lulucopter opened the pull request:
# Dan

Looking at your repository, this application sparked some ideas on adding in new features to spice up your program. I have hence created `Dan`, a **sp00ky** :ghost:  variation of your `Duke` program.

Dan is an interactive, *boosted* to-do list, planner all in one, freeing up your headspace so you can keep your focus on the now without worry of forgetting anything important.

> "Having a clear mind and a clear space allows you to think and act with purpose" - Erika Oppenheimer ([source](https://www.erikaoppenheimer.com/blog/two-must-dos-for-better-organization))

The application is 
- easy to use
- easy to learn
- multiplies your productivity
- And the best part, ***SUPER FAST***

To get started, just follow the steps below:
1. Download it from [here](https://github.com/lulucopter/ip/releases)
2. Go to the directory which you had installed it in
3. Open the application using `java -jar dan.jar` (make sure you have java on your computer!)
4. And you are ready to get started! 😄 

There are the commands that are currently available:
- [x] `bye` (exits the program)
- [x] `list` (lists out the tasks that are currently in your list)
- [x] `mark <task number>` (mark a task as completed)
- [x] `unmark <task number>` (unmark a completed task)
- [x] `todo <task description>` (creates a to-do task)
- [x] `deadline <task description> /by <datetime>` (creates a deadline)
- [x] `event <task description> /at <datetime>` (creates an event)
- [x] `delete <task number>` (deletes a task)
- [x] `find <keyword>` (finds a task)


Here is the `main` method in case you can't find it
```java
public static void main(String[] args) {
        Application.launch(Dan.class, args);
}
```

Commit:
Add Gradle support

lulucopter committed:
Implemented basic greeting, echo and exit cases

lulucopter committed:
Added List, Created Task class

lulucopter committed:
Amend Task toString method

lulucopter committed:
Rename list and it's methods to Tasks

lulucopter committed:
Implement mark as done feature

lulucopter committed:
Create new Task type: Deadline class

lulucopter committed:
Create new Task type: ToDo class

lulucopter committed:
Create new Task type: Event class

lulucopter committed:
Create addTask method. Now also shows total tasks

lulucopter committed:
Implement todo, event, deadline Task type into addTask method

lulucopter committed:
Create DanException class

lulucopter committed:
Implement DanException to account for IO errors in task creation

lulucopter committed:
Fix minor bugs with related to DanException

lulucopter committed:
Add deleteTask method

lulucopter committed:
Implement delete action

lulucopter committed:
Enhance readability

lulucopter committed:
Enhance readability

lulucopter committed:
Create class to handle creation and reading of data storage file

lulucopter committed:
Create method to represent task as text data

lulucopter committed:
Fix syntax issues on toDataString method

lulucopter committed:
Implement method to write to data file

lulucopter committed:
Rename program to Dan

lulucopter committed:
Add save functionality

lulucopter committed:
Test Data file

lulucopter committed:
Merge branch 'branch-Level-7'

lulucopter committed:
Change dateString from String type to LocalDateTime type

lulucopter committed:
Add data folder to .gitignore

lulucopter committed:
Remove data from tracking

lulucopter committed:
Fix index out of bounds bug with mark/unmark methods

lulucopter committed:
OOP-ify List and its methods into a TaskList

lulucopter committed:
Create Parser class and abstract away main logic into run method

lulucopter committed:
Merge branch 'add-gradle-support'

lulucopter committed:
Package classes into a single package: Dan

lulucopter committed:
Rename package to comply with convention

Donovan9617 commented on src/main/java/Dan/Dan.java:
```diff
@@ -0,0 +1,53 @@
+import java.util.Scanner;
+
+
+public class Dan {
```
Perhaps you forgot java docs for this class and its methods?

lulucopter replied:
Thanks for spotting this!

Donovan9617 commented on src/main/java/Dan/Parser.java:
```diff
@@ -0,0 +1,57 @@
+    public boolean Parse(String input) {
+        String action = input.split(" ")[0];
+        try {
+            switch (action) {
+            case "bye":
+                Ui.sayonara();
+                return true;
+
+            case "list":
+                tasks.showTasks();
+                break;
+
+            case "mark":
+                tasks.markTask(Integer.parseInt(input.split(" ")[1]));
+                break;
+
+            case "unmark":
+                tasks.unMarkTask(Integer.parseInt(input.split(" ")[1]));
+                break;
+
+            case "delete":
+                tasks.deleteTask(Integer.parseInt(input.split(" ")[1]));
+                break;
+
+            case "todo":
+                //fall through
+            case "deadline":
+                //fall through
+            case "event":
+                tasks.addTask(input);
+                break;
+
+            default:
+                throw new DanException("I don't really understand what do you mean by that...");
```
Nice job following the coding standards for the switch clause!

zsiggg replied:
It's good that the case clauses are not indented, but I think there shouldn't be an empty line between cases as well? Maybe its for clarity, but from what I read in the style guide it seems like there shouldn't be the extra lines.

lulucopter replied:
Thanks for pointing this out! Will take note of this.

Donovan9617 commented on src/main/java/Dan/TaskList.java:
```diff
@@ -0,0 +1,103 @@
+    }
+
+    public void addTask(String input) throws DanException {
+        String description;
+        String dateString;
```
Did you forget access modifiers here?

lulucopter replied:
These are actually part of addTask method, and not an attribute of the class

Donovan9617 commented on src/main/java/Dan/Task.java:
```diff
@@ -0,0 +1,28 @@
+
+    @Override
+    public String toString() {
+        return String.format("[%s] %s",getStatusIcon(), this.description);
```
There should be a space gap between , and getStatusIcon()

lulucopter replied:
Missed this out. Thank you!

Donovan9617 reviewed (commented):
Nice job following coding standards. Overall LGTM!

lulucopter committed:
Add JUnit tests for TaskList methods

lulucopter committed:
Add JUnit tests for exception thrown in TaskList methods

lulucopter committed:
Add JUnit test for Parser class: parse method

zsiggg commented on src/main/java/Dan/Parser.java:
```diff
@@ -0,0 +1,57 @@
+        try {
+            switch (action) {
+            case "bye":
+                Ui.sayonara();
```
Sayonara isn't in English, so maybe we could change this to `Ui.bye()` instead?

lulucopter replied:
Had forgotten about this! Thanks for pointing it out

zsiggg commented on src/main/java/Dan/Parser.java:
```diff
@@ -0,0 +1,57 @@
+                break;
+
+            case "unmark":
+                tasks.unMarkTask(Integer.parseInt(input.split(" ")[1]));
```
Since method names should be in camelCase, perhaps this should be `tasks.unmarkTask()` instead? Because I see unmark as one word.

lulucopter replied:
You are right. Thought that unMark were 2 different words

zsiggg commented on src/main/java/Dan/TaskList.java:
```diff
@@ -0,0 +1,103 @@
+            tasks.add(new ToDo(description));
+
+        } else if (input.startsWith("deadline")) {
+            String[] temp = input.replace("deadline","").strip().split("/by");
```
Because we are using `temp` in multiple places within this `else if` block, might want to name it more informatively? Perhaps like `inputArr`?

lulucopter replied:
Thanks for the suggestion

zsiggg commented on src/main/java/Dan/TaskListReader.java:
```diff
@@ -0,0 +1,74 @@
+    /* Stores data separated by <SEPARATOR> in the format with no whitespace:
+    eg. SEPARATOR = "|"
+    Tasktype{T,D,E}|Task.isDone{1,0}|Task.Name|(optional)Deadline String
+     */
```
Maybe these could be made as JavaDoc comments for `TaskListReader`?

zsiggg commented on src/main/java/Dan/TaskListReader.java:
```diff
@@ -0,0 +1,74 @@
+        Files.write(filePath, taskList.getTasks().stream()
+                .map(x -> x.toDataString(SEPARATOR))
+                .collect(Collectors.toList())
+        );
```
Nice use of streams 🔥

zsiggg reviewed (commented):
Overall, coding standard mostly adhered to. LGTM after the few comments are resolved, and as the previous reviewer mentioned, the JavaDocs for public classes/methods are made. :)

lulucopter committed:
Add Javadocs to Dan class

lulucopter committed:
Add Javadocs to TaskListReader.java

lulucopter committed:
Add Javadocs for DanException class

lulucopter committed:
Add javadocs for the Task class

lulucopter committed:
Add javadocs for the ToDo, Event and Deadline task types

lulucopter committed:
Add javadocs to Parser and Ui classes

lulucopter committed:
Tweak code to comply with coding standard

lulucopter committed:
Add Find task functionality to the chatbot

lulucopter committed:
Merge Level-9 branch into master

lulucopter committed:
Merge A-JavaDoc branch into master

lulucopter committed:
Repackage classes into relevant packages

lulucopter committed:
Add import statements to match the new package structure

lulucopter committed:
Amend test files to new package structure

lulucopter committed:
Rephrase Javadoc header comments

lulucopter committed:
Merge branch 'master' of https://github.com/lulucopter/ip

lulucopter committed:
Amend code to improve readability

lulucopter committed:
Implement checkstyle plugin

lulucopter committed:
Update checkstyle to v10.2

lulucopter committed:
Set program flow for JavaFxml

lulucopter committed:
Refactor code to fit JavaFxml

lulucopter committed:
Implement JavaFxml GUI

lulucopter committed:
Package main driver and GUI into dan.app

lulucopter committed:
Align gradle dependencies with new package structure

lulucopter committed:
Merge branch 'branch-level-10'

lulucopter committed:
A-Assertions: add assert statements to the code

Currently assumptions are made on invariants as the code logic
is processed.

Assumptions may be made about internal invariants, etc., but these
may not hold in unforeseen circumstances, which will cause the code
and logic to break, and potentially cause irreversible damage.

Let's include assert statements into the code logic.

Assert statements allow us to liberally ensure that these
assumptions are hold as the code is processed, ensuring that
unforeseen circumstances are accounted for,
safeguarding the program and data from damage.

Assert statements are preferred to other methods, as the runtime
will terminate the program immediately, safeguarding the program
from continuing.

Refer to this S/O discussion on why assert is preferred
"https://stackoverflow.com/questions/11468899/any-benefit-
of-using-assert-instead-of-using-a-simple-if"

lulucopter committed:
Merge pull request #2 from lulucopter/branch-A-Assertions

Branch A Assertions

lulucopter committed:
Rename Dan to dan

lulucopter committed:
Remove warning suppressions

lulucopter committed:
DanException: Refactor errors into static methods

Currently, DanException is thrown with custom messages when certain
error conditions are met. As the code base grows, the conditions
which may trigger these errors become hard to read, and the reader
must reverse engineer the code to figure out what causes these
errors.

Extracting these common behavior into static methods with magic
strings help to improve the readability of the code, in line with
the DRY and SLAP principles.

These makes the code more declarative and readable, while at the
same time, reduce human error by declaring error messages into a
single static final String.

lulucopter committed:
addTask: extract logic into 3 separate methods

Currently, the addTask fully handles the "parsing" of the user
input to choose the appropriate task to be added for the addTask
method.

This can complicate readability of the code, and create confusion.
The parser method should fully handle which type of task should be
added.

Extracting the logic into 3 separate methods enhances the code
quality and readability, as well as helping to reduce arrowheads in
the code, complying with the SLAP principle.

This can set the code design up for further changes to be made in
its appropriate method, allowing for easier extension of features
and changes in the future.

lulucopter committed:
Merge pull request #3 from lulucopter/branch-A-CodeQuality

Merge Branch-A-code-quality

lulucopter committed:
Add varargs feature using streams for findTask, markTask,
unmarkTask and deleteTask methods.

lulucopter committed:
Merge pull request #4 from lulucopter/branch-A-Streams

Merge branch-A-Streams

lulucopter committed:
Set up CI using github actions

lulucopter committed:
C-BetterSearch: Add Pattern Regex matching to find function

Currently, the find method already searches for partial substring
matches (oops haha).

Let's take it a step further, and empower the user to use pattern
regex substring matching to give users more flexibility in looking
for what they need in the find function.

More details on the Pattern Regex available can be found here:
https://docs.oracle.com/javase/8/docs/api/java/util/regex/Pattern.html

lulucopter committed:
Merge branch 'branch-BCD-Extension'

lulucopter committed:
Update welcome message and fix minor bugs

lulucopter committed:
Recolor GUI to increase aesthetics

lulucopter committed:
Merge branch 'branch-A-BetterGui'

lulucopter committed:
Upload product screenshot to docs

lulucopter committed:
Add background to GUI

lulucopter committed:
A-UserGuide: Update README.md to show the User Guide

lulucopter committed:
Update README.md

lulucopter committed:
Update README.md

lulucopter committed:
Clean up dead code

lulucopter committed:
Fix checkstyle issues
